src/campaigns/tasks/services/aiCampaignManager.service.ts

```diff
@@ -0,0 +1,214 @@
+            line,
+            parseError,
+          })
+        }
+      }
+
+      return progressData
+    } catch (error) {
+      this.logger.error('Failed to get progress stream', error)
+      throw new BadGatewayException('Failed to get progress stream')
+    }
+  }
+
+  async waitForCompletion(
+    sessionId: string,
+    maxWaitTimeMs = 300000, // 5 minutes default
+    pollIntervalMs = 5000, // 5 seconds default
+  ): Promise<ProgressStreamData> {
+    const startTime = Date.now()
+
+    while (Date.now() - startTime < maxWaitTimeMs) {
+      const progressData = await this.getProgressStream(sessionId)
+      const latestProgress = progressData[progressData.length - 1]
+
+      if (latestProgress?.status === 'completed') {
+        return latestProgress
+      }
+
+      if (latestProgress?.status === 'failed') {
+        throw new BadGatewayException('Campaign plan generation failed')
+      }
+
+      await this.sleep(pollIntervalMs)
+    }
+
+    throw new BadGatewayException('Campaign plan generation timed out')
+  }
+
+  private async fetchFromApi<T>(
+    endpoint: string,
+    options: {
+      method?: Methods
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      data?: any
+      headers?: Record<string, string>
+    } = {},
+  ): Promise<AxiosResponse<T>> {
+    try {
+      const { method = Methods.GET, data, headers = {} } = options
+      const url = `${this.apiBaseUrl}${endpoint}`
+
+      let requestData = data
+      const config = {
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+          ...headers,
+        },
+      }
+
+      if (method === Methods.POST && data && typeof data === 'object') {
+        const params = new URLSearchParams()
+        Object.keys(data).forEach((key) => {
+          if (data[key] !== null && data[key] !== undefined) {
+            params.append(key, String(data[key]))
+          }
+        })
+        requestData = params.toString()
+        this.logger.debug('URL-encoded request data:', requestData)
+      }
+
+      let response: AxiosResponse<T>
+      switch (method) {
+        case Methods.POST:
+          response = await lastValueFrom(
+            this.httpService.post(url, requestData, config),
+          )
+          break
+        case Methods.PUT:
+          response = await lastValueFrom(
+            this.httpService.put(url, data, config),
+          )
+          break
+        case Methods.DELETE:
+          response = await lastValueFrom(this.httpService.delete(url, config))
+          break
+        default:
+          response = await lastValueFrom(this.httpService.get(url, config))
+      }
+
+      return response
+    } catch (error) {
+      this.logger.error(
+        `Failed to ${options.method || Methods.GET} ${endpoint}`,
+        error,
+      )
+      throw new BadGatewayException(
+        'Failed to communicate with AI Campaign Manager API',
+      )
+    }
+  }
```

Comment:
This really is complete overkill.

Simply using the `HttpService`'s verb methods directly in our services should be fine.

We _really_ don't want to abstract out fetch and paint ourselves into a corner like the similar abstraction in the webapp has done.

Reply:
Furthermore, an abstraction like this, if it _were_ something we agreed upon to be abstracted, should be a shared abstraction across services, not just one service.